From e2f3e4439ada412f097eaee787606419f3f45288 Mon Sep 17 00:00:00 2001
From: Christopher Albert <albert@tugraz.at>
Date: Fri, 14 Nov 2025 11:39:49 +0100
Subject: [PATCH] fortran: clarify spec-stmt ordering error

* parse.cc (parse_executable): Reject declaration/OpenMP spec statements
  with "in the execution part" diagnostic when seen after executable code.
* testsuite/gfortran.dg/spec_statement_in_exec.f90: New test.

PR fortran/32365

Signed-off-by: Christopher Albert <albert@tugraz.at>
---
 gcc/fortran/parse.cc                          |  9 ++++
 .../gfortran.dg/spec_statement_in_exec.f90    | 44 +++++++++++++++++++
 2 files changed, 53 insertions(+)
 create mode 100644 gcc/testsuite/gfortran.dg/spec_statement_in_exec.f90

diff --git a/gcc/fortran/parse.cc b/gcc/fortran/parse.cc
index f987f464023..96382512884 100644
--- a/gcc/fortran/parse.cc
+++ b/gcc/fortran/parse.cc
@@ -7132,6 +7132,15 @@ loop:
 	  accept_statement (st);
 	  goto done;
 
+	/* Specification statements cannot appear after executable statements.  */
+	case_decl:
+	case_omp_decl:
+	  gfc_error ("%s statement at %C after the first executable statement",
+		     gfc_ascii_statement (st));
+	  reject_statement ();
+	  st = next_statement ();
+	  continue;
+
 	default:
 	  break;
 	}
diff --git a/gcc/testsuite/gfortran.dg/spec_statement_in_exec.f90 b/gcc/testsuite/gfortran.dg/spec_statement_in_exec.f90
new file mode 100644
index 00000000000..268304bc049
--- /dev/null
+++ b/gcc/testsuite/gfortran.dg/spec_statement_in_exec.f90
@@ -0,0 +1,44 @@
+! { dg-do compile }
+! { dg-options "-fopenmp" }
+! Test improved error messages for specification statements in executable section
+! PR fortran/32365 - Better error message for specification statement in executable section
+
+subroutine test_spec_in_exec
+  implicit none
+  integer :: i
+
+  ! First executable statement
+  i = 1
+
+  ! Test key specification statement types
+  integer :: j                     ! { dg-error "data declaration statement cannot appear after an executable statement" }
+  real :: x                       ! { dg-error "data declaration statement cannot appear after an executable statement" }
+  complex :: z                    ! { dg-error "data declaration statement cannot appear after an executable statement" }
+  logical :: flag                  ! { dg-error "data declaration statement cannot appear after an executable statement" }
+  character(len=20) :: name       ! { dg-error "data declaration statement cannot appear after an executable statement" }
+  double precision :: d           ! { dg-error "data declaration statement cannot appear after an executable statement" }
+  common /myblock/ i              ! { dg-error "COMMON statement cannot appear after an executable statement" }
+  equivalence (i, i)              ! { dg-error "EQUIVALENCE statement cannot appear after an executable statement" }
+  namelist /nml/ i                ! { dg-error "NAMELIST statement cannot appear after an executable statement" }
+  data i / 2 /                    ! { dg-error "DATA statement cannot appear after an executable statement" }
+!$omp threadprivate(i)             ! { dg-error "THREADPRIVATE statement cannot appear after an executable statement" }
+!$omp declare target (i)           ! { dg-error "DECLARE TARGET statement cannot appear after an executable statement" }
+
+end subroutine test_spec_in_exec
+
+! Test that correct order still works fine
+subroutine test_correct_order
+  implicit none
+  integer :: i, j
+  real :: x
+
+  ! All specification statements first
+  common /myblock/ i, j
+  data i, j / 1, 2 /
+!$omp threadprivate(i)
+
+  ! Then executable statements
+  i = 4
+  x = 3.14
+
+end subroutine test_correct_order
-- 
2.51.2

